File: Problems/Algorithms/50. Pow(x, n)/pow.py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        if x == 0:
            return 0
        if n == 0:
            return 1

        if n < 0:
            x = 1/x
            n = -n

        return self.myPow(x*x, n//2) if n % 2 == 0 else x * self.myPow(x*x, n//2)

File: Problems/Algorithms/50. Pow(x, n)/pow.rs
impl Solution {
    pub fn my_pow(mut x: f64, n: i32) -> f64 {
        match n {
            0 => 1.0,
            -2147483648 => Self::my_pow(x*x, -2147483648/2),
            m if m < 0 => 1.0 / Self::my_pow(x, -m),
            _ => {
                if n % 2 == 0 {
                    Self::my_pow(x*x, n/2)
                } else {
                    x * Self::my_pow(x*x, n/2)
                }
            },
        }
    }
}

